File: src/laolx/parser/Token.java
/*
 * The MIT License
 *
 * Copyright 2017 kwpfalzer.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package laolx.parser;

import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Stream;

/**
 * A token.
 *
 * @author kwpfalzer
 */
public class Token {

    public Token(Location location, Code code, String text) {
        this.location = location;
        this.code = code;
        this.text = text;
    }

    public final Location location;
    public final Code code;
    public final String text;

    /**
     * Force class to load and static initialize.
     */
    public static void initialize() {
    }

    /**
     * Check if text matches a keyword.
     *
     * @param text text to check.
     * @return Code of matching keyword or null if not a keyword.
     */
    public static Code isKeyword(String text) {
        return KEYWORDS.get(text);
    }

    /**
     * Check is text matches a keyword.
     *
     * @param text text to check.
     * @param ref set to keyword code iff text is a keyword; else do nothing.
     * @return true if text matches keyword (and ref[0] updated); else false;
     */
    public static boolean isKeyword(String text, Code ref[]) {
        final Code kwrd = isKeyword(text);
        if (Objects.nonNull(kwrd)) {
            assert 0 < ref.length;
            ref[0] = kwrd;
        }
        return Objects.nonNull(kwrd);
    }

    /**
     * Token codes.
     */
    public static enum Code {
        EOF,
        EOLN,
        WS, //whitespace (not including EOF or EOLN)
        LINE_COMMENT,
        BLOCK_COMMENT,
        IDENT,
        DQSTRING, //double-quoted string
        SQSTRING, //single-quoted string
        SYMBOL, //e.g. :foobar
        REGEXP, //%r{...}
        //
        // K_xxx are keywords
        K_CASE("case"),
        K_CATCH("catch"),
        K_CLASS("class"),
        K_DEFAULT("default"),
        K_ELSE("else"),
        K_ELSIF("elsif"),
        K_EXTENDS("extends"),
        K_FLOAT("float"),
        K_FOR("for"),
        K_IF("if"),
        K_IMPLEMENTS("implements"),
        K_INT("int"),
        K_INTERFACE("interface"),
        K_NIL("nil"),
        K_PRIVATE("private"),
        K_PROTECTED("protected"),
        K_PUBLIC("public"),
        K_STATIC("static"),
        K_STRING("string"),
        K_THROW("throw"),
        K_UNLESS("unless"),
        K_UNTIL("until"),
        K_WHEN("when"),
        K_WHILE("while"),
        //
        // S_xxx are symbols
        S_TIDLE("~"),
        S_EXCLAMATION("!"),
        S_AT("@"),
        S_POUND("#"),
        S_DOLLAR("$"),
        S_PCNT("%"),
        S_PCNTEQ("%="),
        S_WORDS("%w{"), //start list of words
        S_SYMBOLS("%s{"), //start list of symbols
        S_CARET("^"),
        S_AND("&"),
        S_ANDEQ("&="),
        S_AND2("&&"),
        S_AND2EQ("&&="),
        S_STAR("*"),
        S_STAREQ("*="),
        S_STAR2("**"),
        S_STAR2EQ("**="),
        S_LPAREN("("),
        S_RPAREN(")"),
        S_MINUS("-"),
        S_PLUS("+"),
        S_PLUS2("++"),
        S_EQ("="),
        S_PLUSEQ("+="),
        S_MINUSEQ("-="),
        S_LCURLY("{"),
        S_RCURLY("}"),
        S_LBRACK("["),
        S_RBRACK("]"),
        S_OR("|"),
        S_OREQ("|="),
        S_OR2("||"),
        S_OR2EQ("||="),
        S_BSLASH("\\"),
        S_COLON(":"),
        S_COLON2("::"),
        S_SEMI(";"),
        S_DQOUTE("\""),
        S_SQUOTE("'"),
        S_LT("<"),
        S_LTEQ("<="),
        S_LT2("<<"),
        S_LT2EQ("<<="),
        S_GT(">"),
        S_GTEQ(">="),
        S_GT2(">>"),
        S_GT2EQ(">>="),
        S_COMMA(","),
        S_DOT("."),
        S_QMARK("?"),
        S_DIV("/"),
        S_BACKTIC(",");

        private Code() {
            this(null);
        }

        private Code(String text) {
            this.text = text;
        }

        final String text;
    }

    public static final Map<String, Code> KEYWORDS = new HashMap<>();
    public static final Map<Character, Code[]> SYMBOLS = new HashMap<>();

    static {
        initKeywords();
        initSymbols();
    }

    /**
     * Initialize KEYWORDS to Code by String/keyword.
     */
    private static void initKeywords() {
        Stream.of(Code.class.getEnumConstants())
                .filter(code -> code.toString().startsWith("K_"))
                .forEach((Code code) -> {
                    KEYWORDS.put(code.text, code);
                });
    }

    /**
     * Initialize SYMBOLS to array Code[] by 1st symbol char. Code[] is sorted
     * longest to shortest *by text.length); i.e., to find longest matching.
     */
    private static void initSymbols() {
        Map<Character, List<Code>> codesByChar = new HashMap<>();
        Stream.of(Code.class.getEnumConstants())
                .filter(code -> code.toString().startsWith("S_"))
                .forEach((Code code) -> {
                    final Character first = code.text.charAt(0);
                    if (!codesByChar.containsKey(first)) {
                        codesByChar.put(first, new LinkedList<>());
                    }
                    codesByChar.get(first).add(code);
                });
        codesByChar.entrySet().stream().forEach(e -> {
            Code[] value = e
                    .getValue()
                    .stream()
                    .sorted((Code o1, Code o2) -> {
                        final int len1 = o1.text.length(), len2 = o2.text.length();
                        return (len1 < len2) ? 1 : ((len1 == len2) ? 0 : -1);
                    })
                    .toArray(Code[]::new);
            SYMBOLS.put(e.getKey(), value);
        });
    }

}

File: laolx2cxx/parser/token.cxx
/*
 * The MIT License
 *
 * Copyright 2017 kpfalzer.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include "parser/token.hxx"

laolx::String Location::toString() const {
    laolx::String loc;
    if (!filename.empty()) {
        loc += filename + ":";
    }
    loc += ":";
    loc += linenum;
    loc += ":";
    loc += colnum;
    return loc;
}

const laolx::String Token::EMPTY = "";

const Token::CodeByString Token::stKeywords({
    // K_xxx are keywords
    {"assert", Token::K_ASSERT},
    {"case", Token::K_CASE},
    {"catch", Token::K_CATCH},
    {"class", Token::K_CLASS},
    {"const", Token::K_CONST},
    {"default", Token::K_DEFAULT},
    {"else", Token::K_ELSE},
    {"elsif", Token::K_ELSIF},
    {"extends", Token::K_EXTENDS},
    {"extern", Token::K_EXTERN},
    {"float", Token::K_FLOAT},
    {"for", Token::K_FOR},
    {"if", Token::K_IF},
    {"implements", Token::K_IMPLEMENTS},
    {"include", Token::K_INCLUDE},
    {"int", Token::K_INT},
    {"interface", Token::K_INTERFACE},
    {"namespace", Token::K_NAMESPACE},
    {"new", Token::K_NEW},
    {"nil", Token::K_NIL},
    {"private", Token::K_PRIVATE},
    {"protected", Token::K_PROTECTED},
    {"public", Token::K_PUBLIC},
    {"static", Token::K_STATIC},
    {"string", Token::K_STRING},
    {"throw", Token::K_THROW},
    {"unless", Token::K_UNLESS},
    {"until", Token::K_UNTIL},
    {"when", Token::K_WHEN},
    {"while", Token::K_WHILE}
});


const Token::CodeByString Token::stSymbols({
    {"~", Token::S_TIDLE},
    {"!", Token::S_EXCLAMATION},
    {"@", Token::S_AT},
    {"#", Token::S_POUND},
    {"$", Token::S_DOLLAR},
    {"%", Token::S_PCNT},
    {"^", Token::S_PCNTEQ},
    {"%w{", Token::S_WORDS},
    {"%s{", Token::S_SYMBOLS},
    {"^", Token::S_CARET},
    {"&", Token::S_AND},
    {"&=", Token::S_ANDEQ},
    {"&&", Token::S_AND2},
    {"&&=", Token::S_AND2EQ},
    {"*", Token::S_STAR},
    {"*=", Token::S_STAREQ},
    {"**", Token::S_STAR2},
    {"**=", Token::S_STAR2EQ},
    {"(", Token::S_LPAREN},
    {")", Token::S_RPAREN},
    {"-", Token::S_MINUS},
    {"+", Token::S_PLUS},
    {"++", Token::S_PLUS2},
    {"=", Token::S_EQ},
    {"=>", Token::S_EQGT},
    {"+=", Token::S_PLUSEQ},
    {"-=", Token::S_MINUSEQ},
    {"{", Token::S_LCURLY},
    {"}", Token::S_RCURLY},
    {"[", Token::S_LBRACK},
    {"]", Token::S_RBRACK},
    {"|", Token::S_OR},
    {"|=", Token::S_OREQ},
    {"||", Token::S_OR2},
    {"||=", Token::S_OR2EQ},
    {"\\", Token::S_BSLASH},
    {":", Token::S_COLON},
    {"::", Token::S_COLON2},
    {";", Token::S_SEMI},
    {"\"", Token::S_DQOUTE},
    {"'", Token::S_SQUOTE},
    {"<", Token::S_LT},
    {"<=", Token::S_LTEQ},
    {"<<", Token::S_LT2},
    {"<<=", Token::S_LT2EQ},
    {">", Token::S_GT},
    {">=", Token::S_GTEQ},
    {">>", Token::S_GT2},
    {">>=", Token::S_GT2EQ},
    {",", Token::S_COMMA},
    {".", Token::S_DOT},
    {"..", Token::S_DOT2},
    {"...", Token::S_DOT3},
    {"?", Token::S_QMARK},
    {"/", Token::S_DIV},
    {"`", Token::S_BACKTIC},
});

Token::SymbolsByChar Token::stSymbolsByChar;

bool Token::stInited = Token::init();

bool Token::init() {
    stSymbols.each([](const CodeByString::value_type & kv) {
        char ch = kv.first.at(0);
        stSymbolsByChar[ch] << kv;
    });
    //sort by descending length
    stSymbolsByChar.each([](const SymbolsByChar::value_type & kv) {
        auto vals = const_cast<SymbolsByChar::mapped_type&> (kv.second);
        vals.sort([](auto& a, auto& b) {
            return a.first.length() > b.first.length();
        });
    });
    return true;
}

bool Token::isKeyword(const laolx::String& text, Code& code) {
    const bool isKeyword = stKeywords.hasKey(text);
    code = isKeyword ? stKeywords[text] : INVALID;
    return isKeyword;
}

Token::~Token() {
}
